src/core/service.c

```diff
@@ -3385,7 +3387,13 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         else
                 assert_not_reached();
 
-        if (s->main_pid == pid) {
+        /* Services with ExitType=cgroup ignore the main PID for purposes of exit status */
+        if (s->exit_type == SERVICE_EXIT_CGROUP && s->main_pid == pid) {
+                service_unwatch_main_pid(s);
+                s->main_pid_known = false;
```

Comment:
The exec_fd_event_source should be cleaned up still in this case, i.e the first part of the if block below should run in that case too

Reply:
I think service_load_pid_file() and so on should also still run. i.e. we should still track the main pid as good as we can in this mode. Knowing what the main pid is is one thing, considering the service down whenever the main pid is gone is another.